File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        vector<int> nodes;
        while (head){
            nodes.push_back(head->val);
            head = head->next;
        }
        int i = 0;
        int ans = INT_MIN;
        int j = nodes.size()-1;
        
        while (i<j){
            ans = max(nodes[i] + nodes[j], ans);
            i++;
            j--;
        }
        return ans;
    }
};

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow = head
        fast = head
        ans = 0
        
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
            
        prev = None
        curr = slow
        
        while curr:
            curr.next,prev,curr = prev,curr,curr.next
            
            
        start = head
        
        while prev:
            ans = max(ans,start.val + prev.val)
            prev = prev.next
            start = start.next
            
            
        return ans
        
        return ans